app/backend/src/couchers/rate_limit.py

```diff
@@ -0,0 +1,214 @@
+            .where(Conversation.created >= now() - RATE_LIMIT_INTERVAL)
+        )
+        .mappings()
+        .all()
+    )
+
+
+def _get_user_events_in_past_time_interval(session, user_id: int, action: RateLimitAction) -> list[dict]:
+    """Get all events for the user in the last rate limit interval for the mod email."""
+    match action:
+        case RateLimitAction.host_request:
+            return _get_user_host_requests_in_past_time_interval(session, user_id)
+        case RateLimitAction.friend_request:
+            return _get_user_friend_requests_in_past_time_interval(session, user_id)
+        case RateLimitAction.chat_initiation:
+            return _get_user_initiated_chats_in_past_time_interval(session, user_id)
+        case _:
+            raise ValueError(f"Unknown rate limit action: {action}")
+
+
+def _save_rate_limit_violation(
+    session: "Session", user_id: int, action: RateLimitAction, hard_limit: bool
+) -> RateLimitViolation:
+    """Save a rate limit violation to the database and return it."""
+    violation = RateLimitViolation(
+        user_id=user_id,
+        action=action,
+        hard_limit=hard_limit,
+    )
+    session.add(violation)
+    session.flush()
+    return violation
+
+
+def _user_rate_limit_violations_in_past_time_interval(
+    session: "Session", user_id: int, action: RateLimitAction, hard_limit: bool
+) -> int:
+    """Return the number of user rate limit violations in the last rate limit interval."""
+    return session.execute(
+        select(func.count())
+        .select_from(RateLimitViolation)
+        .where(RateLimitViolation.user_id == user_id)
+        .where(RateLimitViolation.action == action)
+        .where(RateLimitViolation.created >= now() - RATE_LIMIT_INTERVAL)
+        .where(RateLimitViolation.hard_limit == hard_limit)
+    ).scalar_one()
+
+
+def _get_rate_limit_for_action(action: RateLimitAction) -> tuple[int, int]:
+    match action:
+        case RateLimitAction.host_request:
+            return HOST_REQUEST_WARNING_LIMIT, HOST_REQUEST_HARD_LIMIT
+        case RateLimitAction.friend_request:
+            return FRIEND_REQUEST_WARNING_LIMIT, FRIEND_REQUEST_HARD_LIMIT
+        case RateLimitAction.chat_initiation:
+            return CHAT_INITIATION_WARNING_LIMIT, CHAT_INITIATION_HARD_LIMIT
+        case _:
+            raise ValueError(f"Unknown rate limit action: {action}")
+
+
+def _has_reached_rate_limit(
+    session: "Session", user_id: int, action: RateLimitAction, count_last_interval: int
+) -> bool:
+    warning_limit, hard_limit = _get_rate_limit_for_action(action=action)
+    for limit, is_hard_limit in [(warning_limit, False), (hard_limit, True)]:
+        if count_last_interval >= limit:
+            if (
+                _user_rate_limit_violations_in_past_time_interval(
+                    session=session, user_id=user_id, action=action, hard_limit=is_hard_limit
+                )
+                == 0
+            ):
```

Comment:
Please change `_user_rate_limit_violations_in_past_time_interval` to be `_user_has_violated_rate_limit_in_past_time_interval` if this is only used to check existence.